docs/proposals/receiving_metadata.md

```diff
@@ -0,0 +1,49 @@
+
+- On average, metadata (a combination of `HELP`, `TYPE`, `UNIT` and `METRIC_NAME`) is ~55 bytes uncompressed.
+- at GL, on an instance with about 2.6M active series, we hold ~1241 unique metrics in total.
+- with that, we can assume that on a worst-case scenario the metadata set for that instance is ~68 kilobytes uncompressed.
+- by default, this data is only propagated once every minute (aligning with the default scrape interval), but this can be adjusted.
+- Finally, what this gives us is a baseline worst-case scenario formula for the data to store per tenant: `~68KB * Replication Factor * # of Instances`. Keeping in mind that typically, there's a very high overlap of metadata across instances, and we plan to deduplicate in the ingesters.
+
+### Write Path
+
+1. Store the metadata directly from the distributors into a cache (e.g. Memcached)
+
+Since metadata is received all at once, we could directly store into an external cache using the tenant ID as a key, and still, avoid a read-modify-write.  However, a very common use case of Cortex is to have multiple Prometheus sending data for the same tenant ID. This complicates things, as it adds a need to have an intermediary merging phase and thus making a read-modify-write inevitable.
+
+2. Keep metadata in memory within the distributors
+
+Similarly to what we do with sample data, we can keep the metadata in-memory in the ingestors and apply similar semantics. I propose to use the tenant ID as a hash key, distribute it to the ingesters (taking into account the replication factor), using a hash map to keep a set of the metadata across all instances for a single tenant, and implement a configurable time-based purge process to deal with metadata churn. Given, we need to ensure fair-use we also propose implementing limits for both the number of metadata entries we can receive and the size of entry(s).
+
+### Read Path
+
+In my eyes, the read path seems to only have one option. At the moment of writing, Cortex uses a [`DummyTargetRetriever`](https://github.com/cortexproject/cortex/blob/master/pkg/querier/dummy.go#L11-L20) as a way to signal that these API endpoints are not implemented. We'd need to modify the Prometheus interface to support a `Context` and extract the tenant ID from there. Then, use the tenant ID to query the ingesters for the data, deduplicate it and serve it.
+
+## Conclusions
+
+I conclude that solution #2 is ideal for this work on the write path. It allows us to use similar semantics to samples, thus reducing operational complexity, and lays a groundwork for when we start receiving metadata alongside samples.
+
+There's one last piece to address: Allowing metadata to survive rolling restarts. Option #1 handles this well, given the aim would be to use an external cache such as Memcached. Option #2 lacks this, as it does not include any plans to persist this data. I'm open to opinions here, but I can see a world where we allow this data to be flushed/recovered on shutdown/startup.
```

Comment:
Given metadata is written from Prometheus to Cortex at every scrape interval and given we rollout 1 ingester at a time, this could be a non-problem, unless for scrape intervals which are very high (ie. > 5 minutes).

Reply:
I agree, I still wanted to note it down as it is a downside to option #1. This might result in a bad experience for a given tenant where temporarily you would not see any metadata, but then again - this is similar to what happens in Prometheus as metadata does not survive restarts.

Reply:
In Cortex, if you will have a replication factor > 1, metadata will survive a single ingester restart. However, contrary to series, given we don't plan to implement metadata transfer on ingesters shutdown at this stage, you may loose metadata if 3 ingesters rollout quicker than the scrape interval. If this will be a real problem, we could add "metadata transfer" to the already existing "chunks transfer" mechanism on ingesters shutdown.

Reply:
Also, metadata is sent every 1min if I read it right, it shouldn't be a problem as the rollout per ingester takes longer than that.

Reply:
Sounds good, thank you both for the info. I take it, I don't need a follow-up action, for now, right? We can decide on whenever we want to support a transfer later on.

Reply:
Yes, just note that because Prometheus sends the metadata every minute, its fine. And that we don't need this high level of consistency (out of sync for upto a min). And that if in the future if its required, we can add transfers, etc.